rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala

```diff
@@ -191,27 +192,45 @@ trait StreamTInstances2 {
       def zero[A](state: Stream[A]): Boolean = state.isEmpty
     }
 
-  implicit def streamTSync[F[_]](implicit F0: Sync[F]): Sync[StreamT[F, ?]] =
+  implicit def streamTSync[F[_]](
+      implicit F0: Sync[F],
+      AL0: Alternative[StreamT[F, ?]]
+  ): Sync[StreamT[F, ?]] =
     new StreamTSync[F]() {
-      implicit val F = F0
+      implicit val F  = F0
+      implicit val AL = AL0
     }
 }
 
 private trait StreamTSync[F[_]] extends Sync[StreamT[F, ?]] with StreamTMonadError[F, Throwable] {
   implicit def F: Sync[F]
+  implicit def AL: Alternative[StreamT[F, ?]]
 
   import cats.effect.ExitCase
 
-  private[this] implicit def unlift[A](v: StreamT[F, A]): F[A] = F.map(StreamT.run(v))(_.head)
-
   def bracketCase[A, B](acquire: StreamT[F, A])(use: A => StreamT[F, B])(
       release: (A, ExitCase[Throwable]) => StreamT[F, Unit]
   ): StreamT[F, B] =
-    StreamT.liftF(
-      F.bracketCase(acquire)(use.andThen(unlift(_)))((a, b) => release(a, b))
+    StreamT.fromStream(
+      F.bracketCase[Stream[A], Stream[B]](StreamT.run(acquire)) {
+        case Stream.Empty => F.pure[Stream[B]](Stream.empty)
+        case cons: Cons[A] =>
+          StreamT
+            .run(cons.map(use).fold(StreamT.empty[F, B])(AL.combineK))
```

Comment:
We must not be using `.run` in here (or anywhere except the end of the world, really), because it forces the computation - and part of the point in having a StreamT is to have it lazy.

Instead of running the entire computation, we should be 'running just the first step, adapting it as needed, and returning a computation that will do the same in its next step'.
The method to 'run a single step is `StreamT#next`.

See the implementation of combineK https://github.com/rchain/rchain/blob/e319d584b7287e7cb7ee7da71f182c4fb36e1dd6/rholang/src/main/scala/coop/rchain/rholang/interpreter/matcher/StreamT.scala#L95

Reply:
kudos for delegating the Sync to the underlying effect though :+1:

Reply:
I mean especially the fact that it's a `MonadError[Throwable`